modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -40,108 +41,180 @@
+     * when the document has no source but it's required.
      */
-    protected void copyMetadata(IndexRequest index, SearchHit doc) {
-        index.parent(fieldValue(doc, ParentFieldMapper.NAME));
-        copyRouting(index, doc);
-        // Comes back as a Long but needs to be a string
-        Long timestamp = fieldValue(doc, TimestampFieldMapper.NAME);
-        if (timestamp != null) {
-            index.timestamp(timestamp.toString());
-        }
-        Long ttl = fieldValue(doc, TTLFieldMapper.NAME);
-        if (ttl != null) {
-            index.ttl(ttl);
+    protected boolean accept(SearchHit doc) {
+        if (doc.hasSource()) {
+            /*
+             * Either the document didn't store _source or we didn't fetch it for some reason. Since we don't allow the user to
+             * change the "fields" part of the search request it is unlikely that we got here because we didn't fetch _source.
+             * Thus the error message assumes that it wasn't stored.
+             */
+            throw new IllegalArgumentException("[" + doc.index() + "][" + doc.type() + "][" + doc.id() + "] didn't store _source");
         }
+        return true;
+    }
+
+    /**
+     * Build the {@link RequestWrapper} for a single search hit. This shouldn't handle
+     * metadata or scripting. That will be handled by copyMetadata and
+     * apply functions that can be overridden.
+     */
+    protected abstract RequestWrapper<?> buildRequest(SearchHit doc);
+
+    /**
+     * Apply various changes depending of the Reindex action and  the search hit,
+     * from copying search hit metadata (parent, routing, etc) to potentially
+     * transforming the {@link RequestWrapper} completely.
+     */
+    protected RequestWrapper<?> apply(RequestWrapper<?> request, SearchHit doc) {
+        return scriptApplier.apply(copyMetadata(request, doc), doc);
+    }
+
+    /**
+     * Copies the metadata from a hit to the request.
+     */
+    protected RequestWrapper<?> copyMetadata(RequestWrapper<?> request, SearchHit doc) {
+        copyParent(request, fieldValue(doc, ParentFieldMapper.NAME));
+        copyRouting(request, fieldValue(doc, RoutingFieldMapper.NAME));
+        copyTimestamp(request, (Long) fieldValue(doc, TimestampFieldMapper.NAME));
+        copyTtl(request, fieldValue(doc, TTLFieldMapper.NAME));
+        return request;
+    }
+
+    /**
+     * Copy the index from a search hit to the request.
+     */
+    protected void copyIndex(RequestWrapper<?> request, String index) {
+        request.setIndex(index);
     }
 
     /**
-     * Part of copyMetadata but called out individual for easy overwriting.
+     * Copy the type from a search hit to the request.
      */
-    protected void copyRouting(IndexRequest index, SearchHit doc) {
-        index.routing(fieldValue(doc, RoutingFieldMapper.NAME));
+    protected void copyType(RequestWrapper<?> request, String type) {
+        request.setType(type);
```

Comment:
I wonder if these methods are important any more now that you have the RequestWrapper? Like maybe just call the RequestWrapper method directly?